File: android/src/main/java/com/reactnativedynamsoftbarcodescanner/DynamsoftBarcodeScannerViewManager.java
package com.reactnativedynamsoftbarcodescanner;

import android.util.Base64;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;

import com.dynamsoft.dbr.BarcodeReader;
import com.dynamsoft.dbr.BarcodeReaderException;
import com.dynamsoft.dbr.DBRLicenseVerificationListener;
import com.dynamsoft.dbr.EnumConflictMode;
import com.dynamsoft.dbr.ImageData;
import com.dynamsoft.dbr.TextResult;
import com.dynamsoft.dbr.TextResultListener;
import com.dynamsoft.dce.CameraEnhancer;
import com.dynamsoft.dce.CameraEnhancerException;
import com.dynamsoft.dce.DCECameraView;
import com.dynamsoft.dce.DCELicenseVerificationListener;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.uimanager.SimpleViewManager;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.annotations.ReactProp;


public class DynamsoftBarcodeScannerViewManager extends SimpleViewManager<DCECameraView> implements LifecycleEventListener {
    public static final String REACT_CLASS = "DynamsoftBarcodeScannerView";
    private CameraEnhancer mCameraEnhancer;
    private DCECameraView mCameraView;
    private BarcodeReader reader = null;
    private ThemedReactContext context;
    private String dbrLicense = null;
    private String template = null;
    private String dceLicense = "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ==";
    private Boolean flashOn = false;
    private Boolean isScanning = false;

    @Override
    @NonNull
    public String getName() {
        return REACT_CLASS;
    }

    @Override
    @NonNull
    public DCECameraView createViewInstance(ThemedReactContext reactContext) {
        context = reactContext;
        reactContext.addLifecycleEventListener(this);
        mCameraView = new DCECameraView(reactContext.getBaseContext());
        return mCameraView;
    }

    @ReactProp(name = "dceLicense")
    public void setDceLicense(View view, String license) {
        dceLicense = license;
    }

    @ReactProp(name = "template")
    public void setTemplate(View view, String template) {
        this.template = template;
        updateTemplate();
    }

    @ReactProp(name = "flashOn")
    public void setFlashOn(View view, Boolean on) {
        flashOn = on;
        toggleFlash();
    }

    private void updateSettings(){
        toggleFlash();
        updateTemplate();
    }

    private void toggleFlash(){
        if (mCameraEnhancer!=null){
            try {
                if (flashOn == true){
                    mCameraEnhancer.turnOnTorch();
                }else{
                    mCameraEnhancer.turnOffTorch();
                }
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
    }

    private void updateTemplate(){
        if (reader!=null && template!=null){
            try {
                reader.initRuntimeSettingsWithString(template, EnumConflictMode.CM_OVERWRITE);
            } catch (BarcodeReaderException e) {
                e.printStackTrace();
            }
        }
    }

    @ReactProp(name = "dbrLicense")
    public void setDbrLicense(View view, String license) {
        dbrLicense = license;
    }

    @ReactProp(name = "cameraID")
    public void setCameraID(View view, String id) {
        if (mCameraEnhancer!=null){
            try {
                mCameraEnhancer.selectCamera(id);
            } catch (CameraEnhancerException  e) {
                e.printStackTrace();
            }
        }
    }

    @ReactProp(name = "isScanning")
    public void setIsScanning(View view, Boolean isScanning) {
        this.isScanning = isScanning;
        if (isScanning) {
            initIfNeeded();
            updateSettings();
            try {
                mCameraEnhancer.open();
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }else{
            if (mCameraEnhancer!=null){
                try {
                    mCameraEnhancer.close();
                } catch (CameraEnhancerException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    private void initIfNeeded(){
        if (reader == null){
            initDBR();
            initDCE(context);
            bindDBRandDCE();
        }
    }

    private void initDBR()  {
        String license;
        if (dbrLicense != null){
            license = dceLicense;
        }else{
            license = "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ==";
        }
        BarcodeReader.initLicense(license, new DBRLicenseVerificationListener() {
            @Override
            public void DBRLicenseVerificationCallback(boolean isSuccess, Exception error) {
                if(!isSuccess){
                    error.printStackTrace();
                }
            }
        });
        try {
            reader = new BarcodeReader();
        } catch (BarcodeReaderException e) {
            e.printStackTrace();
        }
    }

    private DCECameraView initDCE(ThemedReactContext reactContext){
        CameraEnhancer.initLicense(dceLicense, new DCELicenseVerificationListener() {
            @Override
            public void DCELicenseVerificationCallback(boolean isSuccess, Exception error) {
                if(!isSuccess){
                    error.printStackTrace();
                }
            }
        });

        mCameraEnhancer = new CameraEnhancer(reactContext.getBaseContext());
        onCameraOpened();
        mCameraEnhancer.setCameraView(mCameraView);
        mCameraView.setOverlayVisible(true);
        return mCameraView;
    }

    private void bindDBRandDCE(){
        TextResultListener mTextResultListener = new TextResultListener() {
            // Obtain the recognized barcode results and display.
            @Override
            public void textResultCallback(int id, ImageData imageData, TextResult[] textResults) {
                onScanned(textResults);
            }
        };
        reader.setTextResultListener(mTextResultListener);
        // Bind the Camera Enhancer instance to the Barcode Reader instance.
        // The mCameraEnhancer is the instance of the Dynamsoft Camera Enhancer.
        // The Barcode Reader will use this instance to take control of the camera and acquire frames from the camera to start the barcode decoding process.
        reader.setCameraEnhancer(mCameraEnhancer);
        // Start the barcode scanning thread.
        reader.startScanning();
    }

    public void onScanned(TextResult[] textResults) {
        WritableArray results = Arguments.createArray();
        for (TextResult tr:textResults){
            WritableMap map = Arguments.createMap();
            map.putString("barcodeText", tr.barcodeText);
            map.putString("barcodeFormat", tr.barcodeFormatString);
            map.putString("barcodeBytesBase64", Base64.encodeToString(tr.barcodeBytes, Base64.DEFAULT));
            results.pushMap(map);
        }
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("onScanned",results);
    }

    public void onCameraOpened() {
        WritableMap map = Arguments.createMap();
        WritableArray cameras = Arguments.createArray();
        for (String cameraID : mCameraEnhancer.getAllCameras()){
            cameras.pushString(cameraID);
        }
        map.putArray("cameras",cameras);
        map.putString("selectedCamera",mCameraEnhancer.getSelectedCamera());
        Log.d("DBR","camera opened");
        context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class).emit("onCameraOpened",map);
    }

    @Override
    public void onHostResume() {
        if (mCameraEnhancer!=null){
            if (this.isScanning){
                try {
                    mCameraEnhancer.open();
                    reader.startScanning();
                } catch (CameraEnhancerException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    @Override
    public void onHostPause() {
        if (mCameraEnhancer!=null){
            try {
                reader.stopScanning();
                mCameraEnhancer.close();
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void onHostDestroy() {
        if (reader!=null){
            reader.destroy();
            reader=null;
            mCameraEnhancer=null;
        }
        context.removeLifecycleEventListener(this);
    }
}

File: ios/DynamsoftBarcodeScannerViewManager.swift
import DynamsoftCameraEnhancer
import DynamsoftBarcodeReader
import React

@objc(DynamsoftBarcodeScannerViewManager)
class DynamsoftBarcodeScannerViewManager: RCTViewManager  {

  override func view() -> (DynamsoftBarcodeScannerView) {
      let view = DynamsoftBarcodeScannerView()
      view.setBridge(bridge: self.bridge)
      return view
  }
}



class DynamsoftBarcodeScannerView : UIView, DBRLicenseVerificationListener, DCELicenseVerificationListener, DBRTextResultListener {

    @objc var dceLicense: String = "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ=="
    @objc var dbrLicense: String = "DLS2eyJoYW5kc2hha2VDb2RlIjoiMjAwMDAxLTE2NDk4Mjk3OTI2MzUiLCJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSIsInNlc3Npb25QYXNzd29yZCI6IndTcGR6Vm05WDJrcEQ5YUoifQ=="
    var dce:DynamsoftCameraEnhancer! = nil
    var barcodeReader:DynamsoftBarcodeReader! = nil
    var dceView:DCECameraView! = nil
    var bridge:RCTBridge! = nil

    func setBridge(bridge:RCTBridge){
        self.bridge = bridge
    }
    
    @objc var isScanning: Bool = false {
        didSet {
            if isScanning
            {
                if (barcodeReader == nil){
                    configurationDBR()
                    configurationDCE()
                    updateSettings()
                }else{
                    barcodeReader.startScanning()
                    dce.resume()
                }
            }else{
                if dce != nil {
                    barcodeReader.stopScanning()
                    dce.pause()
                }
            }
        }
    }
    
    @objc var flashOn: Bool = false {
        didSet {
            toggleFlash()
        }
    }
    
    @objc var cameraID: String = "" {
        didSet {
            if dce != nil {
                if cameraID != "" {
                    var error: NSError? = NSError()
                    dce.selectCamera(cameraID, error: &error)
                }
            }
        }
    }
    
    @objc var template: String = "" {
        didSet {
            updateTemplate()
        }
    }
    
    func updateSettings(){
        toggleFlash()
        updateTemplate()
    }
    
    func toggleFlash(){
        if dce != nil {
            if flashOn
            {
                dce.turnOnTorch()
            }else{
                dce.turnOffTorch()
            }
        }
    }
    
    func updateTemplate(){
        if barcodeReader != nil {
            if (template != ""){
                try? barcodeReader.initRuntimeSettingsWithString(template, conflictMode: EnumConflictMode.overwrite)
            }else{
                try? barcodeReader.resetRuntimeSettings()
            }
        }
    }
    
    func configurationDBR() {
        DynamsoftBarcodeReader.initLicense(dbrLicense, verificationDelegate: self)
        barcodeReader = DynamsoftBarcodeReader.init()
    }
    
    func dbrLicenseVerificationCallback(_ isSuccess: Bool, error: Error?) {
        let err = error as NSError?
        if(err != nil){
            print("Server DBR license verify failed")
        }
    }
    
        
    func configurationDCE() {
        // Initialize a camera view for previewing video.
        dceView = DCECameraView.init(frame: self.bounds)
       
        self.addSubview(dceView)
        dceView.overlayVisible = true
        DynamsoftCameraEnhancer.initLicense(dceLicense, verificationDelegate: self)
        dce = DynamsoftCameraEnhancer.init(view: dceView)
        dce.open()
        onCameraOpened()
        bindDCEtoDBR()
    }

    func bindDCEtoDBR(){
        // Bind Camera Enhancer to the Barcode Reader.
        // The Barcode Reader will get video frame from the Camera Enhancer
        barcodeReader.setCameraEnhancer(dce)

        // Set text result call back to get barcode results.
        barcodeReader.setDBRTextResultListener(self)

        // Start the barcode decoding thread.
        barcodeReader.startScanning()
    }
    
    
    func dceLicenseVerificationCallback(_ isSuccess: Bool, error: Error?) {
        let err = error as NSError?
        if(err != nil){
            print("Server DCE license verify failed")
        }
    }
    
    // Obtain the recognized barcode results from the textResultCallback and display the results
    func textResultCallback(_ frameId: Int, imageData: iImageData, results: [iTextResult]?) {
        let count = results?.count ?? 0
        let array = NSMutableArray()
        for index in 0..<count {
            let tr = results![index]
            let result = NSMutableDictionary()
            result["barcodeText"] = tr.barcodeText
            result["barcodeFormat"] = tr.barcodeFormatString
            result["barcodeBytesBase64"] = tr.barcodeBytes?.base64EncodedString()
            array.add(result)
        }
        bridge.eventDispatcher().sendDeviceEvent(withName: "onScanned", body: array)
    }
    
    public func onCameraOpened(){
        let info = NSMutableDictionary()
        info["selectedCamera"] = dce.getSelectedCamera()
        print(dce.getSelectedCamera())
        let array = NSMutableArray()
        for cameraID in dce.getAllCameras(){
            array.add(cameraID)
        }
        info["cameras"] = array
        bridge.eventDispatcher().sendDeviceEvent(withName: "onCameraOpened", body: info)
    }
}
